index.js

```diff
@@ -55,6 +58,77 @@ app.get("/favicon.ico", (_req, res) => {
 app.all("/graphql", (req, res) => res.redirect("/"))
 
 app.use(bodyParser.json())
+
+function parse_args() {
+  return "( ... )"
+}
+
+function trace(res, span) {
+  span.addTags({
+    "http.status_code": res.statusCode,
+  })
+  span.finish()
+}
+
+app.use((req, res, next) => {
+  if (req.method === "POST") {
+    const tracer = new Tracer({ service: "metaphysics" })
+    const span = tracer.startSpan("metaphysics.query")
+    const query = req.body.query.replace(/(\()([^\)]*)(\))/g, parse_args)
+    span.addTags({
+      resource: query,
+      type: "web",
+      "span.kind": "server",
+      "http.method": req.method,
+      "http.url": req.url,
+    })
+
+    assign(req, { span })
+
+    res.on("finish", () => trace(res, span))
+    res.on("close", () => trace(res, span))
+  }
+  next()
+})
+
+function wrapResolve(typeName, fieldName, resolver) {
+  return function (root, options, request) {
+    const parentSpan = request.span
+    const span = parentSpan.tracer().startSpan("metaphysics.resolver." + typeName + "." + fieldName,
+      { childOf: parentSpan.context() })
+    span.addTags({
+      resource: typeName + ": " + fieldName,
+      type: "web",
+      "span.kind": "server",
+    })
+
+    assign(request, { span })
+    const result = resolver.apply(this, arguments);
+    assign(request, { span: parentSpan })
+
+    if (result instanceof Promise) {
+      return result.finally(function () {
+        span.finish()
+      })
+    }
+
+    span.finish()
+    return result;
+  };
+}
+
+// Walk the schema and for all object type fields with resolvers wrap them in our tracing resolver.
+forIn(schema._typeMap, function (value, key) {
+  const typeName = key
+  if (has(value, "_fields")) {
+    forIn(value._fields, function (field, fieldName) {
+      if (has(field, "resolve") && field.resolve instanceof Function) {
+        assign(field, { resolve: wrapResolve(typeName, fieldName, field.resolve) })
```

Comment:
jslint was getting mad at me for "assigning to a function param".

Reply:
Ah I see. If we’re going to mutate anyways, then I’d vote for disabling the rule for that line rather than jumping to a bunch of extra hoops that make the code slightly less clear.